### PR TITLE
Option to encode cookie values

### PR DESCRIPTION
**Pull request by psiinon**
This changes the default behaviour, but I think for the better.
Help PR will follow.

<details><summary>Active Scan Input Vectors Options dialog</summary>
<p>

Screenshot 2023-11-21 at 12 16 09

</p>
</details>

### CODE REVIEW

**Comment by psiinon** [comment]
Looking at the failing tests now..

**Comment by kingthorin** [comment]
🤦‍♂️ spotless

**Comment by psiinon** [comment]
ugh, did that before, so that must be a problem with the properties file :/

**Comment by psiinon** [comment]
Help PR: https://github.com/zaproxy/zap-core-help/pull/536

**Comment by psiinon** [comment]
All tests passing now 😉

**Comment by kingthorin** [comment]
Hmm see the help PR before my comments here

**Comment by psiinon** [comment]
Yeah, looks like this is a misunderstanding.
Its a global option, encode cookie values or dont encode them.
I think the previous way of always encoding them was probably wrong.

I'm not sure attacking both encoded and unencoded makes a lot of sence (from a number of requests point of view), but I'm willing to be convinced otherwise.
I would rather have an initialisation step when we inject non alphanumberic values into a cookie and see if they are reflected encoded or not.
But that would be a future change.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/VariantFactory.java [hunk 1]
```diff
@@ -135,7 +135,9 @@ public List<Variant> createVariants(ScannerParam scanOptions, HttpMessage messag
         }
 
         if ((targets & ScannerParam.TARGET_COOKIE) != 0) {
-            listVariant.add(new VariantCookie());
+            VariantCookie vc = new VariantCookie();
+            vc.setEncodeCookieValues(scanOptions.isEncodeCookieValues());
```
Comment (kingthorin) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/VariantFactory.java:139]: Would it be "better" to have a constructor that accepts a boolean and then have the existing constructor deprecated and use "false"? Then setup of the Variant isn't a two step process?

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java [hunk 2]
```diff
@@ -121,7 +123,7 @@ public List<NameValuePair> getParamList() {
     @Override
     public String setParameter(
             HttpMessage msg, NameValuePair originalPair, String name, String value) {
-        return setParameter(msg, originalPair, name, value, false);
+        return setParameter(msg, originalPair, name, value, !encodeCookieValues);
```
Comment (kingthorin) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java:126]: This negative logic bothers me. Above we set the default as false so won't !false be true? (In the GUI won't the checkbox be unchecked if the default is false?)

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/VariantFactory.java (hunk 1 above)
Comment (psiinon) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/VariantFactory.java:139]: I thought of that, but ended up copying the way VariantJSONQuery did it above. 
Wasnt sure what the implications would be, but a deprecated constructor would solve that..

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java (hunk 2 above)
Comment (psiinon) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java:126]: Will be changing this as part of a change suggested by @thc202 in irc 😉

**Review by kingthorin — APPROVED** [review]
LGTM

File: zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java [hunk 3]
```diff
@@ -190,4 +198,13 @@ private String setParameter(
 
         return name + "=" + escapedValue;
     }
+
+    /**
+     * Sets whether or not the cookie values should be encoded.
+     *
+     * @since 2.15.0
+     */
+    public void setEncodeCookieValues(boolean encodeCookieValues) {
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/VariantCookie.java:207]: Does it worth keeping this?

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java [hunk 4]
```diff
@@ -415,6 +426,7 @@ public void initParam(ScannerParam param) {
                         getChkInjectableHeaders().isEnabled()
                                 && getChkInjectableHeaders().isSelected());
         this.getChkInjectableCookie().setSelected((targets & ScannerParam.TARGET_COOKIE) != 0);
+        this.getChkEncodeCookieValues().setSelected(param.isEncodeCookieValues());
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java:429]: Should check if `getChkInjectableCookie()` is enabled as well.

**Comment by psiinon** [comment]
Fixed those..

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**